src/main/java/net/rubygrapefruit/platform/internal/jni/AbstractFileEventFunctions.java

```diff
@@ -46,25 +50,47 @@ public AbstractWatcherBuilder(FileWatcherCallback callback) {
      * Configures a new watcher using a builder. Call {@link AbstractWatcherBuilder#start()} to
      * actually start the {@link FileWatcher}.
      */
-    public abstract AbstractWatcherBuilder newWatcher(FileWatcherCallback callback);
+    public abstract AbstractWatcherBuilder newWatcher(BlockingQueue<FileWatchEvent> queue);
 
     protected static class NativeFileWatcherCallback {
-        private final FileWatcherCallback delegate;
 
-        public NativeFileWatcherCallback(FileWatcherCallback delegate) {
-            this.delegate = delegate;
+        private final BlockingQueue<FileWatchEvent> eventQueue;
+
+        public NativeFileWatcherCallback(BlockingQueue<FileWatchEvent> eventQueue) {
+            this.eventQueue = eventQueue;
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void pathChanged(int type, String path) {
-            delegate.pathChanged(FileWatcherCallback.Type.values()[type], path);
+        public void pathChanged(int typeIndex, String path) throws InterruptedException {
+            FileWatchEvent.Type type = FileWatchEvent.Type.values()[typeIndex];
+            if (type == FileWatchEvent.Type.OVERFLOWED) {
+                signalOverflow(path);
+            } else {
+                queueEvent(new ChangeEvent(type, path), false);
+            }
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void reportError(Throwable ex) {
-            delegate.reportError(ex);
+        public void reportError(Throwable ex) throws InterruptedException {
+            queueEvent(new ErrorEvent(ex), true);
+        }
+
+        private void queueEvent(FileWatchEvent event, boolean deliverOnOverflow) throws InterruptedException {
+            // TODO Make the timeout configurable
+            if (!eventQueue.offer(event, 1, SECONDS)) {
+                NativeLogger.LOGGER.info("Event queue overflow, dropping all events");
+                signalOverflow(null);
+                if (deliverOnOverflow) {
+                    eventQueue.put(event);
+                }
+            }
+        }
+
+        private void signalOverflow(@Nullable String path) throws InterruptedException {
+            eventQueue.clear();
+            eventQueue.put(new ChangeEvent(FileWatchEvent.Type.OVERFLOWED, path));
```

Comment:
I think we should use `offer` without a timeout here and fail if the queue cannot take the single event. Blocking doesn't make sense anyhow, right?

The question is what to do if the queue doesn't accept the overflow event, even when we just cleared it. Throwing an exception from here causes an error to be reported, which probably also can't be put in the queue. We could set a flag here that everything is borked and then
- try to report an error every time an event comes in
- throw an exception when closing the watcher

Reply:
I like the idea to only use `offer()`. I don't like the idea of recording the invalid state and acting differently if it's set; it adds too much complexity.

We are only pushing to it from the processor thread, so it should be empty after `clear()`. The only two exceptions I have in mind is if it's a custom queue with 0 capacity, or if another thread is adding elements to the queue (which it shouldn't).

Neither of these seem like cases we need to handle in any graceful way. So throwing an `IllegalStateException` sounds good to me. WDYT?

Reply:
Throwing an `IllegalStateException` is perfectly fine with me. The problem with that is that the exception will end up in the native code which will then report the illegal state exception as an error, causing another illegal state exception since we still can't add to the queue, right?

Reply:
I thought about this a bit more. The best we can do in this situation I think is to log the event and then ignore the fact that we had such a problem.